### ISSUE TITLE
MSYS2 SDL2: Screen blanked by Windows lock screen

### ISSUE THREAD

**Issue by smbhax** [angband/angband#5497]
After a Windows lock screen is cleared from hiding Angband's SDL2 front-end, Angband's screen will be blank, and can only be refreshed partially; it is necessary to restart the program to restore the full game screen.

MSYS2-compiled latest Angband source running SDL2 front end in Windows 11 Home

1) Run Angband with the SDL2 front-end in Windows
2) Get a Windows User account control lock screen. This may be achieved in various ways, for instance:
-- Win+L to lock the screen
-- Run a program requiring administrator permission, such as running Notepad++ as Administrator, or running an old program like Photoshop 4 that wants to write to the registry
3) Clear the Windows lock screen
4) Return to Angband
== Angband's screen is blank. CTRL+R to refresh will restore segments of the screen, but text in particular remains obscured, and the game is effectively unplayable until it is restarted.

![ang_blanked](https://user-images.githubusercontent.com/91744319/196016707-675cad17-af4d-42e6-a96f-1158f8d457c5.jpg)

Video: https://youtu.be/HoSvVNSILf4?t=5123

- - - - -

SDL discussion of what may be a similar issue

I don't know how applicable this may be to Angband, but I came across a 2016 developer discussion of something like this being caused by Windows Win+L lock screens over their SDL 2.0.4 program; the person who replied mentioned finding a solution for their own version of the issue, involving moving "all SDL graphics calls to one thread and one main loop": https://discourse.libsdl.org/t/renderer-freeze-when-lock-screen-and-login-again/21654/3

- - - - -

Windows workaround for pop-up User account control screens when running programs requiring admin permission:
- Run Windows 11's "Change User Account Control settings" control panel
- Lower the setting slider from 3/4 ("default") to 2/4 ("do not dim my desktop")
= Now User account control pop-ups will not take over the whole screen, and will not corrupt Angband's SDL2 front end display

**Comment by backwardsEric** [comment]
Does changing the font on the window (from the SDL2 front end's menu bar) after redrawing has been broken by the lock screen cause the text areas to reappear?

**Comment by smbhax** [comment]
Yep, toggling the fonts in the main window and subwindows seems to get everything visible again, with no need for the partial restore via CTRL+R.

**Comment by backwardsEric** [comment]
So it sounds like this is the same problem mentioned here, https://stackoverflow.com/questions/50301271/sdl2-texture-cleared-if-i-go-to-windows-lock-screen-and-come-back-or-resize-wind .  I haven't found much in the way of a solution or workaround on Angband's end.  There's https://discourse.libsdl.org/t/renderer-freeze-when-lock-screen-and-login-again/21654 where the workaround is to recreate the persistent textures and the renderer on window enter events.  Another possibility is to recreate the textures (and possibly the renderer) in response to the TERM_XTRA_REACT that CTRL-R (and some other things) triggers.

**Comment by smbhax** [comment]
Not possible to use OpenGL rather than DirectX, then? : )

**Comment by smbhax** [comment]
Come to think of it, given that the problem doesn't occur with the Cygwin/X and WSL Ubuntu front ends, I wonder if those are using OpenGL. Cygwin/X does say something about OpenGL in its opening dialogues.

Eh oh but those two versions also have delayed input and sound response for me in Windows--which is why I'm not using them--so yeah maybe that wouldn't be great. Yeah, if that's what switching to OpenGL would do, please don't switch to OpenGL. ^ _^

**Comment by backwardsEric** [comment]
Given that using OpenGL should go directly to the OpenGL implementation on Windows rather than being routed through X11, the performance could be acceptable.  Though, It looks like there'd have to be substantial changes to main-sdl2.c to route the rendering through OpenGL.

I do have one more question about your environment.  Since this could be an issue with the Windows graphics driver, what discrete or integrated GPU are you using and what's the version of the driver for that hardware?
